src/warp.jl

```diff
@@ -1,51 +1,35 @@
-@inline _getindex(A, v::StaticVector) = A[convert(Tuple, v)...]
-
-warp(img::AbstractArray, args...) = warp(interpolate(img, BSpline(Linear()), OnGrid()), args...)
-
-@inline _dst_type{T<:Colorant,S}(::Type{T}, ::Type{S}) = ccolor(T, S)
-@inline _dst_type{T<:Number,S}(::Type{T}, ::Type{S}) = T
-
-function warp{T,S}(::Type{T}, img::AbstractArray{S}, args...)
-    TCol = _dst_type(T,S)
-    TNorm = eltype(TCol)
-    apad, pad = Interpolations.prefilter(TNorm, TCol, img, typeof(BSpline(Linear())), typeof(OnGrid()))
-    itp = Interpolations.BSplineInterpolation(TNorm, apad, BSpline(Linear()), OnGrid(), pad)
-    warp(itp, args...)
-end
-
-@compat const FloatLike{T<:AbstractFloat} = Union{T,AbstractGray{T}}
-@compat const FloatColorant{T<:AbstractFloat} = Colorant{T}
-
-# The default values used by extrapolation for off-domain points
-@inline _default_fill{T<:FloatLike}(::Type{T}) = convert(T, NaN)
-@inline _default_fill{T<:FloatColorant}(::Type{T}) = nan(T)
-@inline _default_fill{T}(::Type{T}) = zero(T)
-
-warp{T}(img::AbstractInterpolation{T}, tform, fill=_default_fill(T)) = warp(extrapolate(img, fill), tform)
-
 """
-    warp(img, tform) -> imgw
+    warp(img, tform, [indices], [degree = Linear()], [fill = NaN]) -> imgw
 
 Transform the coordinates of `img`, returning a new `imgw` satisfying
-`imgw[x] = img[tform(x)]`. `tform` should be defined using
-CoordinateTransformations.jl.
+`imgw[I] = img[tform(I)]`. This approach is known as backward
+mode warping. The transformation `tform` should be defined using
+[CoordinateTransformations.jl](https://github.com/FugroRoames/CoordinateTransformations.jl).
 
 # Interpolation scheme
 
-At off-grid points, `imgw` is calculated by interpolation. The default
-is linear interpolation, used when `img` is a plain array, and `NaN`
-values are used to indicate locations for which `tform(x)` was outside
-the bounds of the input `img`. For more control over the interpolation
-scheme---and how beyond-the-edge points are handled---pass it in as an
-`AbstractExtrapolation` from Interpolations.jl.
+At off-grid points, `imgw` is calculated by interpolation. The
+degree of the b-spline can be specified with the optional
+parameter `degree`, which can take the values `Linear()` or
+`Constant()`.
+
+The b-spline interpolation is used when `img` is a plain array
+and `img[tform(I)]` is inbound. In the case `tform(I)` maps
+to indices outside the original `img`, the value or extrapolation
+scheme denoted by the optional parameter `fill` (which defaults
+to `NaN` if the element type supports it, and `0` otherwise) is
+used to indicate locations for which `tform(I)` was outside the
+bounds of the input `img`.
+
+For more control over the interpolation scheme --- and how
+beyond-the-edge points are handled --- pass it in as an
+`AbstractInterpolation` or `AbstractExtrapolation` from
+[Interpolations.jl](https://github.com/JuliaMath/Interpolations.jl).
```

Comment:
It's interesting how these packages fit together.

It might be worth noting that `interpolations` isn't the only piece of the puzzle regarding image warping.  In particular, it's never going to deal with proper antialiasing which can be quite important when the warping is severe or you're downsampling by a large factor.

For severe warping and large downsampling things can get quite interesting, as you need full anisotropic filtering, usually coupled with image pyramiding to make it efficient.  Arguably, `warp` could do these things with appropriate options turned on.